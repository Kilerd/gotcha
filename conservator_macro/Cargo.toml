[package]
name = "conservator_macro"
description = "conservator macro"
version = "0.1.9"
edition = "2021"
license = "MIT"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lib]
proc-macro = true

[dependencies]
darling = "0.13.0"
itertools = "0.10.3"
proc-macro2 = "1.0.33"
quote = "1.0.10"
regex = "1.5.4"
strum = { version = "0.23", features = ["derive"] }
syn = "1.0.82"
proc-macro-error = { version = "1", default-features = false }

[dev-dependencies]
trybuild = "1.0"
