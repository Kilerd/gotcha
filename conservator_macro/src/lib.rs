use proc_macro::TokenStream;
use proc_macro_error::{abort, proc_macro_error};

mod authorization;
mod auto;
mod creatable;
mod domain;
mod sql;

#[proc_macro_derive(Domain, attributes(domain))]
#[proc_macro_error]
pub fn derive_domain_fn(input: TokenStream) -> TokenStream {
    let stream2 = proc_macro2::TokenStream::from(input);
    match domain::handler(stream2) {
        Ok(stream) => proc_macro::TokenStream::from(stream),
        Err((span, msg)) => abort! {span, msg},
    }
}

#[proc_macro_derive(Creatable)]
#[proc_macro_error]
pub fn derive_creatable_fn(input: TokenStream) -> TokenStream {
    let stream2 = proc_macro2::TokenStream::from(input);
    let stream1 = proc_macro::TokenStream::from(creatable::handle_creatable(stream2));
    stream1
}

#[proc_macro_attribute]
pub fn auto(_args: TokenStream, input: TokenStream) -> TokenStream {
    let stream2 = proc_macro2::TokenStream::from(input);
    let stream1 = proc_macro::TokenStream::from(auto::handler(stream2));
    stream1
}

#[proc_macro_attribute]
#[proc_macro_error]
pub fn sql(args: TokenStream, input: TokenStream) -> TokenStream {
    let args = proc_macro2::TokenStream::from(args);
    let stream2 = proc_macro2::TokenStream::from(input);
    match sql::handler(args, stream2) {
        Ok(stream) => proc_macro::TokenStream::from(stream),
        Err((span, msg)) => abort! {span, msg},
    }
}
