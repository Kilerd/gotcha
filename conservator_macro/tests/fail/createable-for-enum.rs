use conservator_macro::Creatable;

#[derive(Creatable)]
pub enum NotSupport {
    A,
    B,
}

fn main() {}
