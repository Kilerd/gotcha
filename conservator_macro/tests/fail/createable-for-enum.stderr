error: enum does not support
 --> tests/fail/createable-for-enum.rs:4:1
  |
4 | / pub enum NotSupport {
5 | |     A,
6 | |     B,
7 | | }
  | |_^
